### PR TITLE
ASoC: DPCM fixes

### PR DESCRIPTION
**Pull request by plbossart**
This version is a lot simpler/consistent than previous attempts, let's start reviewing it. 

I need to double-check results with SoundWire/HDaudio in deep-buffer mode, but it's ready for internal reviews.

### CODE REVIEW

**Comment by plbossart** [comment]
sent as upstream RFC - no issues detected in validation so far, apart from alsa-bat settings that need to be updated for the new topologies.

File: sound/soc/soc-pcm.c [hunk 1]
```diff
@@ -2051,34 +2051,53 @@ int dpcm_be_dai_trigger(struct snd_soc_pcm_runtime *fe, int stream,
...
-			if (ret)
+			if (ret) {
+				be->dpcm[stream].be_start--;
 				goto end;
+			}
 
 			be->dpcm[stream].state = SND_SOC_DPCM_STATE_START;
 			break;
 		case SNDRV_PCM_TRIGGER_PAUSE_RELEASE:
 			if ((be->dpcm[stream].state != SND_SOC_DPCM_STATE_PAUSED))
```
Comment (RanderWang) [inline sound/soc/soc-pcm.c]: how about 
````
if ((be->dpcm[stream].state != SND_SOC_DPCM_STATE_PAUSED) || be->dpcm[stream].be_start)
    continue
````
and remove " if (be->dpcm[stream].be_start != 1) "

Comment (plbossart) [reply sound/soc/soc-pcm.c]: @RanderWang your proposal mixes state handling and refcount handling. I think it's better to keep the two parts separate for clarity. The goal is not to have minimalistic code, it's really to make the state management clearer.

Comment (RanderWang) [reply sound/soc/soc-pcm.c]: sure. I don't like 
````
if (be->dpcm[stream].be_start != 1)
````
how about

````
if (be->dpcm[stream].be_start > 1)  // multiple call
      continue;
````

File: sound/soc/soc-pcm.c [hunk 2]
```diff
@@ -1842,7 +1865,7 @@ void dpcm_be_dai_hw_free(struct snd_soc_pcm_runtime *fe, int stream)
 
 		/* only free hw when no longer used - check all FEs */
 		if (!snd_soc_dpcm_can_be_free_stop(fe, be, stream))
-				continue;
+			continue;
```
Comment (bardliao) [inline sound/soc/soc-pcm.c:1868]: This looks right, but maybe belong to a different commit.

File: sound/soc/soc-pcm.c (hunk 1 above)
Comment (bardliao) [reply sound/soc/soc-pcm.c]: @RanderWang IMHO, I like `if (be->dpcm[stream].be_start != 0)` and `if (be->dpcm[stream].be_start != 1)`. which says that the code is only executed when be_start is 0 or 1 exactly.

File: sound/soc/soc-pcm.c (hunk 2 above)
Comment (plbossart) [reply sound/soc/soc-pcm.c:1868]: done.

**Comment by plbossart** [comment]
replaced by PR #3209
